### ISSUE TITLE
Python 3.11 is not yet supported for building

### ISSUE THREAD

**Issue by greschd** [ansys/ansys-tools-protoc-helper#29]
TL;DR for users of this library: use Python <3.11 for _building_ your API library wheels. The resulting wheels can be installed also in 3.11+.

---

Support for Python 3.11 was added in `grpcio-tools` version `1.49.0`. 

Currently, this means that the API libraries can be _built_ only with a Python `<=3.10`. They can still be _installed_ into `3.11+`.

Upgrading the `grpcio-tools` version used here would mean that _only_ `grpc` versions `1.49+` will be supported by the generated wheels.

For now, the best way forward is probably to explicitly require Python <3.11 in this library. Eventually, we will have to drop support for grpc `<1.49`.

@akaszynski @MaxJPRey, opinions?

**Comment by MaxJPRey** [comment]
Hi @greschd .
I agree that for the time being, we could explicitly require Python <3.11 to keep supporting our former versions of `grpcio-tools`.

**Comment by RobPasMue** [comment]
I haven't tried it yet but using 3.11 for build shouldn't be a problem anymore, right @greschd?

**Comment by RobPasMue** [comment]
I think it's time that we only support above 1.49...

**Comment by MaxJPRey** [comment]
That would be great indeed.
If it does not create any conflicts, we should try to proceed with it.

**Comment by greschd** [comment]
Hmm, the PyAnsys metapackage is still stuck on `protobuf==3.20.3` as of the last release. 

This raises the question - can we upgrade grpcio independently, or do we _have to_ stick to "Match the version of protobuf to the version of protoc bundled into grpcio-tools, or at least 3.19.3." (from https://github.com/ansys/ansys-tools-protoc-helper?tab=readme-ov-file#upgrade-plans).

AFAICT, since

"The protoc compiler version used is determined by the grpcio-tools package, not the protobuf dependency. The grpcio-tools==1.20.0 uses protoc==3.7.0."

we'll have to stick to this `grpcio-tools` version until we're able to bump the protobuf version to 4.x.

**Comment by greschd** [comment]
@RobPasMue do you want to do another push to get everyone over the line on protobuf 4.X support?

**Comment by RobPasMue** [comment]
I would say so yeah... we should try to move it. Problem is that due to this limitation we are blocked on building with 3.1o the API packages. And making local development for proto files is becoming a mess for certain teams since we just dropped 3.9... so they can ONLY use 3.10 now. We need to start moving this. I would recommend proposing it in the next PyAnsys Dev meeting.

**Comment by RobPasMue** [comment]
Findings - the following packages are the ones limiting protobuf 4.X

``ansys-dyna-core`` via ``ansys-api-dyna`` --> ~=3.20
``ansys-systemcoupling-core``
``ansys-sherlock-core``
``ansys-edb-core``

By commenting these out from the metapackage, I was able to build it using a condition of ``protobuf>=4``

**Comment by greschd** [comment]
Btw, there's a direct way to check for this:

```
pip install git+https://github.com/ansys/pyansys.git@main#egg=pyansys[all]
pip install protobuf==4.21.1
```

gives an error like 

```
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
googleapis-common-protos 1.63.2 requires protobuf!=3.20.0,!=3.20.1,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0.dev0,>=3.20.2, but you have protobuf 4.21.1 which is incompatible.
google-api-core 2.19.1 requires protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<6.0.0.dev0,>=3.19.5, but you have protobuf 4.21.1 which is incompatible.
ansys-systemcoupling-core 0.6 requires protobuf<4.0.0,>=3.20.1, but you have protobuf 4.21.1 which is incompatible.
ansys-sherlock-core 0.6.0 requires protobuf~=3.20, but you have protobuf 4.21.1 which is incompatible.
ansys-edb-core 0.1.5 requires protobuf<4,>=3.19.3, but you have protobuf 4.21.1 which is incompatible.
ansys-api-systemcoupling 0.1.0 requires protobuf~=3.20, but you have protobuf 4.21.1 which is incompatible.
ansys-api-sherlock 0.1.27 requires protobuf~=3.19, but you have protobuf 4.21.1 which is incompatible.
ansys-api-dyna 0.3.6 requires protobuf~=3.19, but you have protobuf 4.21.1 which is incompatible.
```

**Comment by RobPasMue** [comment]
Yep, thanks! :)

**Comment by RobPasMue** [comment]
I just opened 2 PRs:

- Sherlock repo: https://github.com/ansys/ansys-api-sherlock/pull/56
- DYNA repo: https://github.com/ansys/ansys-api-dyna/pull/13

Once they get merged and released, we would have solved the lowest level issues

System coupling already released a new version of ``ansys-api-systemcoupling`` that allows for protobuf 4.X (see https://github.com/ansys/ansys-api-systemcoupling/pull/7 and released in https://github.com/ansys/ansys-api-systemcoupling/releases/tag/v0.2.0). Next step is that they do a new PySystem Coupling release

Same goes for EDB core. Change is already done, but we are waiting on a release that contains this change.

**Comment by SMoraisAnsys** [comment]
> I just opened 2 PRs:
> 
> * Sherlock repo: [feat: allow for protobuf 4.X and 5.X ansys-api-sherlock#56](https://github.com/ansys/ansys-api-sherlock/pull/56)
> * DYNA repo: [feat: allowing protobuf 4.X and 5.X ansys-api-dyna#13](https://github.com/ansys/ansys-api-dyna/pull/13)
> 
> Once they get merged and released, we would have solved the lowest level issues. Then, both main projects (PyDYNA and PySherlock) would have to update their deps to use these new versions.
> 
> System coupling already released a new version of `ansys-api-systemcoupling` that allows for protobuf 4.X (see [ansys/ansys-api-systemcoupling#7](https://github.com/ansys/ansys-api-systemcoupling/pull/7) and released in https://github.com/ansys/ansys-api-systemcoupling/releases/tag/v0.2.0). Next step is that they do a new PySystem Coupling release
> 
> Same goes for EDB core. Change is already done, but we are waiting on a release that contains this change.

I just notified `pyedb-core` team, they are quite reactive so I hope it should be done soon !

**Comment by RobPasMue** [comment]
Awesome, thanks @SMoraisAnsys!

**Comment by SMoraisAnsys** [comment]
> Awesome, thanks @SMoraisAnsys!

Lastest release of `pyedb-core` allows protobuf 4 see https://github.com/ansys/pyedb-core/blob/c6a8dd7ecafa4a7fbf5fc3c97b004c28fd1c0773/pyproject.toml#L30

**Comment by RobPasMue** [comment]
Thanks @SMoraisAnsys!

**Comment by RobPasMue** [comment]
gRPC autogenerated packages for PyDYNA and PySherlock have been released. We are pending on follow up releases on the main repos of:

- [ ] PySherlock
- [ ] PyDyna
- [ ] PySystem Coupling

**Comment by RobPasMue** [comment]
PyDyna just released a new version with this change - one less to go =)

**Comment by RobPasMue** [comment]
Ah well... PySherlock is not ready yet (but almost!) https://github.com/ansys/pysherlock/pull/371

**Comment by RobPasMue** [comment]
Also @greschd - what if we moved directly to allowing protobuf 5.X? Most packages are there already, only a few are missing.

```sh
ERROR: pip's dependency resolver does not currently take into account all the packages that are installed. This behaviour is the source of the following dependency conflicts.
ansys-additive-core 0.18.1 requires protobuf<5,>=3.20.2, but you have protobuf 5.28.0 which is incompatible.
ansys-api-edb 1.0.6 requires protobuf<5,>=3.19.3, but you have protobuf 5.28.0 which is incompatible.
ansys-api-fluent 0.3.27 requires protobuf<5,>=3.19, but you have protobuf 5.28.0 which is incompatible.
ansys-api-mapdl 0.5.2 requires protobuf<5,>=3.19, but you have protobuf 5.28.0 which is incompatible.
ansys-api-meshing-prime 0.1.3 requires protobuf<5,>=3.19, but you have protobuf 5.28.0 which is incompatible.
ansys-api-systemcoupling 0.1.0 requires protobuf~=3.20, but you have protobuf 5.28.0 which is incompatible.
ansys-api-tools-filetransfer 0.1.0 requires protobuf<5,>=3.19, but you have protobuf 5.28.0 which is incompatible.
ansys-edb-core 0.1.6 requires protobuf<5,>=3.19.3, but you have protobuf 5.28.0 which is incompatible.
ansys-sherlock-core 0.7.0 requires protobuf~=3.20, but you have protobuf 5.28.0 which is incompatible.
ansys-systemcoupling-core 0.6 requires protobuf<4.0.0,>=3.20.1, but you have protobuf 5.28.0 which is incompatible.
Successfully installed protobuf-5.28.0
```

**Comment by greschd** [comment]
> moved directly to allowing protobuf 5

I think we should at least have an intermediate release with 4.x to not delay this further.

**Comment by RobPasMue** [comment]
All good then :) step by step 🏄🏼

**Comment by greschd** [comment]
Generally, I think packages should open up the upper limit as much as possible, or even remove it. That'll make future evolution of this package much easier.

**Comment by RobPasMue** [comment]
I can only agree.


> Generally, I think packages should open up the upper limit as much as possible, or even remove it. That'll make future evolution of this package much easier.

**Comment by iboyd-ansys** [comment]
@RobPasMue I saw your issue earlier and have resolved it with a new 0.7.0 release of PySystemCoupling.

**Comment by RobPasMue** [comment]
Supported - closing
